### PR TITLE
fix addTopicToMirrorMaker

### PR DESCRIPTION
**Pull request by Technoboy-**

### CODE REVIEW

File: uReplicator-Common/src/main/java/com/uber/stream/kafka/mirrormaker/common/utils/PartitionAllocator.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+ */
+public class PartitionAllocator {
+
+    public static List<Integer> allocate(InstanceTopicPartitionHolder currentInstance, List<Integer> partitionAll,
+                                         List<InstanceTopicPartitionHolder> instanceAll) {
+        if (currentInstance == null) {
+            throw new IllegalArgumentException("currentInstance is empty");
+        }
+        if (partitionAll == null || partitionAll.isEmpty()) {
+            throw new IllegalArgumentException("partitionAll is null or mqAll empty");
```
Comment (yangy0000) [inline uReplicator-Common/src/main/java/com/uber/stream/kafka/mirrormaker/common/utils/PartitionAllocator.java]: what does mqAll means?

File: uReplicator-Common/src/main/java/com/uber/stream/kafka/mirrormaker/common/utils/PartitionAllocator.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        }
+        if (instanceAll == null || instanceAll.isEmpty()) {
+            throw new IllegalArgumentException("instanceAll is null or cidAll empty");
+        }
+        if (!instanceAll.contains(currentInstance)) {
+            throw new IllegalArgumentException(String.format("currentInstance : %s not in instanceAll : %s", currentInstance, instanceAll));
+        }
+        final List<Integer> result = new ArrayList<>();
```
Comment (yangy0000) [inline uReplicator-Common/src/main/java/com/uber/stream/kafka/mirrormaker/common/utils/PartitionAllocator.java]: you might want to use ImmutableList from guava

File: uReplicator-Common/src/main/java/com/uber/stream/kafka/mirrormaker/common/utils/PartitionAllocator.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+        int index = instanceAll.indexOf(currentInstance);
+        int mod = partitionAll.size() % instanceAll.size();
+        int averageSize =
```
Comment (yangy0000) [inline uReplicator-Common/src/main/java/com/uber/stream/kafka/mirrormaker/common/utils/PartitionAllocator.java]: add some comments for this logic

File: uReplicator-Common/src/main/java/com/uber/stream/kafka/mirrormaker/common/utils/PartitionAllocator.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        int startIndex = (mod > 0 && index < mod) ? index * averageSize : index * averageSize + mod;
+        int range = Math.min(averageSize, partitionAll.size() - startIndex);
+        for (int i = 0; i < range; i++) {
+            result.add(partitionAll.get((startIndex + i) % partitionAll.size()));
+        }
+        return result;
+    }
+
+    public static String partitionToString(List<Integer> partitions){
+        StringBuilder builder = new StringBuilder(20);
```
Comment (yangy0000) [inline uReplicator-Common/src/main/java/com/uber/stream/kafka/mirrormaker/common/utils/PartitionAllocator.java]: return StringUtils.join(",", partitions);

File: uReplicator-Controller/src/main/java/com/uber/stream/kafka/mirrormaker/controller/core/IdealStateBuilder.java [hunk 5]
```diff
@@ -120,7 +143,7 @@ public static IdealState expandCustomRebalanceModeIdealStateFor(IdealState oldId
         customModeIdealStateBuilder.assignInstanceAndState(Integer.toString(i),
             liveInstance.getInstanceName(),
             "ONLINE");
-        liveInstance.addTopicPartition(new TopicPartition(topicName, i));
```
Comment (yangy0000) [inline uReplicator-Controller/src/main/java/com/uber/stream/kafka/mirrormaker/controller/core/IdealStateBuilder.java]: I don't really understand the logic here, why add only one topic partition to the liveinstance ?

File: uReplicator-Controller/src/main/java/com/uber/stream/kafka/mirrormaker/controller/core/ManagerControllerHelix.java [hunk 6]
```diff
@@ -160,7 +162,7 @@ private void handleRouteAssignmentOnline(String srcCluster, String dstCluster, S
 
     String clusterName = CONTROLLER_WORKER_HELIX_PREFIX + srcCluster + "-" + dstCluster + "-" + routePartition;
     _controllerConf.setHelixClusterName(clusterName);
-    _controllerConf.setEnableSrcKafkaValidation("true");
+    _controllerConf.setEnableSrcKafkaValidation("false");
```
Comment (yangy0000) [inline uReplicator-Controller/src/main/java/com/uber/stream/kafka/mirrormaker/controller/core/ManagerControllerHelix.java]: shall we make this configurable?

File: uReplicator-Common/src/main/java/com/uber/stream/kafka/mirrormaker/common/utils/PartitionAllocator.java (hunk 1 above)
Comment (Technoboy-) [reply uReplicator-Common/src/main/java/com/uber/stream/kafka/mirrormaker/common/utils/PartitionAllocator.java]: sorry, forget to delete.

**Pull request closed without merging**